### PR TITLE
[ENG-11829] Revoke ORCID trusted party access during GDPR delete - Part 1

### PR DESCRIPTION
**Pull request by Vlad0n20**
[//]: # (
    * Before submitting your Pull Request, make sure you've picked the right branch and your code is up-to-date with it.
        * For critical hotfixes, select "master" as the target branch.
        * For bugfixes, improvements and new features, select "develop" as the target branch.
        * If your PR is part of a project/team, select project/team's dedicated feature branch as the target.
    * If you have a JIRA ticket, prefix the ticket number [ENG-*****] to the PR title.
)

## Ticket

[//]: # (Link to a JIRA ticket if available.)

* [ENG-11829](https://openscience.atlassian.net/browse/ENG-11829?atlOrigin=eyJpIjoiZTgxOGUyYjgwYjU4NDdhNWE4OGE2ZmQ3MmFiN2Y3NjQiLCJwIjoiaiJ9)

## Purpose

Revoke ORCID trusted party access during GDPR delete - Part 1

## Changes

[//]: # (Briefly describe or list your changes.)

## Side Effects

[//]: # (Any possible side effects?)

## QE Notes

[//]: # (
    * Any QA testing notes for QE?
        * Make verification statements inspired by your code and what your code touches.
        * What are the areas of risk?
        * Any concerns/considerations/questions that development raised?
    * If you have a JIRA ticket, make sure the ticket also contains the QE notes.
)

## CE Notes

[//]: # (
    * Any server configuration and deployment notes for CE?
        * Is model migration required? Is data migration/backfill/population required?
            * If so, is it reversible? Is there a roll-back plan?
        * Does server settings needs to be updated?
            * If so, have you checked with CE on existing settings for affected servers?
        * Are there any deployment dependencies to other services?
    * If you have a JIRA ticket, make sure the ticket also contains the CE notes.
)

## Documentation

[//]: # (
    * Does any internal or external documentation need to be updated?
        * If the API was versioned, update the developer.osf.io changelog.
        * If changes were made to the API, link the developer.osf.io PR here.
)

[ENG-11829]: https://openscience.atlassian.net/browse/ENG-11829?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: framework/auth/cas.py [hunk 1]
```diff
@@ -253,6 +272,24 @@ def get_profile_url():
 
     return get_client().get_profile_url()
 
+def save_orcid_access_and_refresh_token_to_user(user, orcid_id: str, access_token: str, refresh_token: str):
+    sentry.log_message(
+        f'CAS response ORCID attributes: user=[{user._id}], orcidId=[{orcid_id}], '
+        f'orcidAccessToken=[{"present" if access_token else "missing"}]',
+        level=logging.INFO,
+    )
```
Comment (cslzchen) [inline framework/auth/cas.py:265]: ~~Only log message to sentry if access token or refresh token are missing, level is warning~~

Updated after 4th pass: log error to sentry if access token missing, log warning to sentry if refresh token missing

File: osf/models/user.py [hunk 2]
```diff
@@ -319,6 +320,17 @@ class OSFUser(DirtyFieldsMixin, GuidMixin, BaseModel, AbstractBaseUser, Permissi
     #   },
     #   ...
     # }
+    external_identity_access_token = DateTimeAwareJSONField(default=dict, blank=True)
```
Comment (cslzchen) [inline osf/models/user.py]: `external_identity_tokens`

File: framework/auth/cas.py [hunk 3]
```diff
@@ -275,6 +312,8 @@ def make_response_from_ticket(ticket, service_url):
         user_updates = {}  # serialize updates to user to be applied async
         # user found and authenticated
         if user and action == 'authenticate':
+            access_token = cas_resp.attributes.get('orcidAccessToken', None)
+            refresh_token = cas_resp.attributes.get('orcidRefreshToken', None)
```
Comment (cslzchen) [inline framework/auth/cas.py]: Move into `if external_credential:`

File: framework/auth/cas.py [hunk 4]
```diff
@@ -290,6 +329,13 @@ def make_response_from_ticket(ticket, service_url):
                 user_updates['accepted_terms_of_service'] = timezone.now()
                 print_cas_log(f'CAS TOS consent checked: {user.guids.first()._id}, {user.username}', LogLevel.INFO)
             # if we successfully authenticate and a verification key is present, invalidate it
+            if external_credential and access_token and refresh_token:
+                save_orcid_access_and_refresh_token_to_user(
+                    user,
+                    external_credential['id'],
+                    access_token,
+                    refresh_token,
+                )
```
Comment (cslzchen) [inline framework/auth/cas.py]: No need for this

File: osf/models/user.py [hunk 5]
```diff
@@ -2137,6 +2154,49 @@ def _clear_identifying_information(self):
...
+            orcid_id: token_entry.get('access_token')
+            for orcid_id, token_entry in self.external_identity_access_token.get('ORCID', {}).items()
+            if token_entry.get('access_token')
+        }
+        sentry.log_message(
+            f'[GDPR delete; _clear_identifying_information] user={self._id}: '
+            f'found {len(orcid_tokens)} ORCID token(s) to revoke',
+            level=logging.INFO,
+        )
+        for orcid_id, orcid_token in orcid_tokens.items():
```
Comment (cslzchen) [inline osf/models/user.py]: We need to check if there is an external_identity but there is no matching external_identity_tokens, we need to fail the GDPR delete with relevant error messsage.

File: osf/models/user.py
Comment (cslzchen) [inline osf/models/user.py:2153]: Make sure admin GDPR delete can detect and show the error we raised in here due to ORCiD revoke failure.

File: osf/models/user.py (hunk 5 above)
Comment (cslzchen) [inline osf/models/user.py]: A few corners cases:

* We may have users having multiple ORCiD ID (maybe this has been fixed)?
* For each ORCiD ID, there should be only one pair of access and refresh token.
* According to https://info.orcid.org/ufaqs/how-can-i-revoke-tokens/, we only need to call revoke on either of them

File: osf/models/user.py [hunk 6]
```diff
@@ -2137,6 +2154,49 @@ def _clear_identifying_information(self):
...
+            sentry.log_message(
+                f'[GDPR delete] user={self._id}: revoking ORCID id={orcid_id} '
+                f'via {website_settings.ORCID_OAUTH_REVOKE_URL}',
+                level=logging.INFO,
+            )
+            try:
+                response = requests.post(
+                    website_settings.ORCID_OAUTH_REVOKE_URL,
```
Comment (cslzchen) [inline osf/models/user.py]: this is the revoke URL so I don't think we need the one in CAS

Comment (cslzchen) [reply osf/models/user.py]: Add this one to website settings

File: framework/auth/cas.py [hunk 7]
```diff
@@ -97,6 +101,10 @@ def get_auth_token_revocation_url(self):
         url = furl(self.BASE_URL).add(path=['oauth2', 'revoke'])
         return url.url
 
+    def get_orcid_token_revocation_url(self):
+        url = furl(self.BASE_URL).add(path=['osf', 'orcid', 'revoke'])
+        return url.url
+
```
Comment (cslzchen) [inline framework/auth/cas.py]: This builds a CAS URL which we don't need

File: framework/auth/cas.py [hunk 8]
```diff
@@ -198,6 +206,17 @@ def revoke_tokens(self, payload):
...
+    def revoke_orcid_token(self, orcid_id):
+        url = self.get_orcid_token_revocation_url()
+        headers = {
+            'Authorization': f'Bearer {settings.CAS_ORCID_REVOKE_SHARED_SECRET}',
+        }
+        resp = requests.post(url, json={'orcid_id': orcid_id}, headers=headers)
+        if resp.status_code == 204:
+            return True
+        else:
+            self._handle_error(resp)
```
Comment (cslzchen) [inline framework/auth/cas.py]: This is no longer used/needed

File: website/settings/defaults.py [hunk 9]
```diff
@@ -393,6 +393,7 @@ def parent_dir(path):
 SPAM_SUBMIT_TASK_HARD_TIME_LIMIT = 90
 
 CAS_SERVER_URL = 'http://localhost:8080'
+CAS_ORCID_REVOKE_SHARED_SECRET = os.environ.get('CAS_ORCID_REVOKE_SHARED_SECRET', 'changeme')
```
Comment (cslzchen) [inline website/settings/defaults.py]: ORCID_OAUTH_CLIENT_ID = os.environ.get('ORCID_OAUTH_CLIENT_ID', 'changeme')
ORCID_OAUTH_CLIENT_SECRET = os.environ.get('ORCID_OAUTH_CLIENT_SECRET', 'changeme')
ORCID_OAUTH_REVOKE_URL = os.environ.get('ORCID_OAUTH_REVOKE_URL', 'https://orcid.org/oauth/revoke')

File: osf/models/user.py [hunk 10]
```diff
@@ -2137,6 +2154,49 @@ def _clear_identifying_information(self):
...
+                    data={
+                        'client_id': website_settings.ORCID_OAUTH_CLIENT_ID,
+                        'client_secret': website_settings.ORCID_OAUTH_CLIENT_SECRET,
+                        'token': orcid_token,
+                    },
+                    timeout=5,
```
Comment (cslzchen) [inline osf/models/user.py]: make this a settings as well, and set a default value to 15s

Comment (cslzchen) [inline osf/models/user.py]: Future, we probably want a retry just in case

File: osf/models/user.py [hunk 11]
```diff
@@ -2137,6 +2154,49 @@ def _clear_identifying_information(self):
...
+                )
+                sentry.log_message(
+                    f'[GDPR delete] user={self._id}: ORCID id={orcid_id} revoked, '
+                    f'status_code={response.status_code}, response_text={response.text}, response={response}',
```
Comment (cslzchen) [inline osf/models/user.py]: we probably don't need the full response

File: osf/models/user.py [hunk 12]
```diff
@@ -2137,6 +2154,49 @@ def _clear_identifying_information(self):
...
+                    level=logging.INFO,
+                )
+                response.raise_for_status()
```
Comment (cslzchen) [inline osf/models/user.py]: Add unit tests for success, invalid client_id/secret, invalid token, empty token

File: osf/models/user.py [hunk 13]
```diff
@@ -2137,6 +2154,49 @@ def _clear_identifying_information(self):
...
+            except requests.exceptions.RequestException as e:
+                sentry.log_message(
+                    f'[GDPR delete] Failed to revoke ORCID token for user {self._id}: {e}',
```
Comment (cslzchen) [inline osf/models/user.py]: Add the orcid id just in case user has mulitple

File: osf/models/user.py [hunk 14]
```diff
@@ -2137,6 +2154,49 @@ def _clear_identifying_information(self):
...
+                    level=logging.ERROR,
+                )
+                sentry.log_exception(e)
+                raise UserStateError(
+                    'Unable to revoke this user\'s ORCID access right now because ORCID\'s '
+                    'service could not be reached. Please try the GDPR delete again later.'
```
Comment (cslzchen) [inline osf/models/user.py]: rephrase to a more general message like fail to revoke

**Review by cslzchen — CHANGES_REQUESTED** [review]
1st pass done

File: framework/auth/cas.py [hunk 15]
```diff
@@ -253,6 +257,24 @@ def get_profile_url():
...
+    sentry.log_message(
+        f'CAS response ORCID attributes: user=[{user._id}], orcidId=[{orcid_id}], '
+        f'orcidAccessToken=[{"present" if access_token else "missing"}]',
+        level=logging.WARNING,
+    )
+    if orcid_id and access_token:
+        provider = settings.EXTERNAL_IDENTITY_PROFILE['OrcidProfile']
+        user.external_identity_access_token.setdefault(provider, {})[orcid_id] = {
+            'access_token': access_token,
+            'refresh_token': refresh_token,
```
Comment (cslzchen) [inline framework/auth/cas.py]: Remove `refresh_token` due to CAS limit

File: osf/models/user.py [hunk 16]
```diff
@@ -2137,6 +2154,50 @@ def _clear_identifying_information(self):
...
+                    level=logging.ERROR,
+                )
+                sentry.log_exception(e)
+                raise UserStateError(
+                    'Fail to revoke ORCID\'s service could not be reached'
+                )
+        else:
+            raise UserStateError(
+                'User do not have connected ORCID'
+            )
```
Comment (cslzchen) [inline osf/models/user.py]: Move this to if statement first, 

```
if failure
    raise/return
the rest for success
```

File: osf/models/user.py [hunk 17]
```diff
@@ -2172,7 +2233,9 @@ def _clear_identifying_information(self):
                 account.profile_url = None
                 account.save()
             self.external_accounts.clear()
+
         self.external_identity = {}
+        self.external_identity_access_token = {}
```
Comment (cslzchen) [inline osf/models/user.py]: rename

File: website/settings/defaults.py [hunk 18]
```diff
@@ -393,6 +393,12 @@ def parent_dir(path):
...
```
Comment (cslzchen) [inline website/settings/defaults.py]: Remove this one

**Review by cslzchen — CHANGES_REQUESTED** [review]
2nd pass done ⭐

File: osf/models/user.py [hunk 19]
```diff
@@ -319,6 +320,16 @@ class OSFUser(DirtyFieldsMixin, GuidMixin, BaseModel, AbstractBaseUser, Permissi
     #   },
     #   ...
     # }
+    external_identity_tokens = DateTimeAwareJSONField(default=dict, blank=True)
+    # Format: {
+    #   <external_id_provider>: {
+    #       <external_id>: {
+    #           "access_token" : <token>,
+    #       }
```
Comment (cslzchen) [inline osf/models/user.py:329]: I was able to WAR-Overlay the `apereo & pac4j` library (see [here](https://github.com/CenterForOpenScience/osf-cas/pull/117/changes#r3898312413)) to include refresh token. Please add the refresh token back.

File: osf/models/user.py [hunk 20]
```diff
@@ -2137,6 +2153,50 @@ def _clear_identifying_information(self):
         '''
         This method ensures a user's info is deleted during a GDPR delete
         '''
+        # A user has at most one ORCID identity, so there is at most one token entry to revoke.
+        orcid_id, token_entry = next(iter(self.external_identity_tokens.get('ORCID', {}).items()), (None, None))
```
Comment (cslzchen) [inline osf/models/user.py:2158]: Please write a small python code snippet for CE to run to check if we have any users with more than one ORCiD ID (via cloud request).

File: osf/models/user.py [hunk 21]
```diff
@@ -2137,6 +2153,50 @@ def _clear_identifying_information(self):
...
+        orcid_token = token_entry.get('access_token') if token_entry else None
```
Comment (cslzchen) [inline osf/models/user.py]: Similarly, add `refresh_token` back and make it optional (just in case some user's privacy settings in ORCiD doesn't release refresh token to us.

File: osf/models/user.py [hunk 22]
```diff
@@ -2137,6 +2153,50 @@ def _clear_identifying_information(self):
...
+        sentry.log_message(
+            f'[GDPR delete; _clear_identifying_information] user={self._id}: '
+            f'{"found" if orcid_token else "no"} ORCID token to revoke',
+            level=logging.INFO,
+        )
+        if not (orcid_id and orcid_token):
+            raise UserStateError(
+                'User do not have connected ORCID'
+            )
+        else:
```
Comment (cslzchen) [inline osf/models/user.py]: No need for `else` since error would have been raised

File: osf/migrations/0053_gdpr_delete_and_orcid_revoke.py [hunk 23]
```diff
@@ -0,0 +1,18 @@
+from django.db import migrations
+import osf.utils.datetime_aware_jsonfield
+import osf.utils.fields
+
+
+class Migration(migrations.Migration):
```
Comment (cslzchen) [inline osf/migrations/0053_gdpr_delete_and_orcid_revoke.py:6]: Rename to `0053_gdpr_delete_and_orcid_revoke`

File: framework/auth/cas.py [hunk 24]
```diff
@@ -253,6 +256,22 @@ def get_profile_url():
 
     return get_client().get_profile_url()
 
+def save_orcid_access_token_to_user(user, orcid_id: str, access_token: str):
+    sentry.log_message(
+        f'CAS response ORCID attributes: user=[{user._id}], orcidId=[{orcid_id}], '
+        f'orcidAccessToken=[{"present" if access_token else "missing"}]',
+        level=logging.WARNING,
+    )
```
Comment (cslzchen) [inline framework/auth/cas.py:265]: Future (new ticket), change this to debug log and only log empty access token error after we dev tested on staging1.

File: framework/auth/cas.py [hunk 25]
```diff
@@ -253,6 +256,22 @@ def get_profile_url():
...
+    if orcid_id and access_token:
+        provider = settings.EXTERNAL_IDENTITY_PROFILE['OrcidProfile']
+        user.external_identity_tokens.setdefault(provider, {})[orcid_id] = {
+            'access_token': access_token,
+        }
+        sentry.log_message(
+            f'ORCID token stored on external_identity_tokens: user=[{user._id}], '
+            f'provider_id=[{orcid_id}], access_token=[{"present" if access_token else "missing"}]',
+            level=logging.INFO,
+        )
```
Comment (cslzchen) [inline framework/auth/cas.py:278]: Ditto

File: framework/auth/cas.py [hunk 26]
```diff
@@ -297,7 +316,13 @@ def make_response_from_ticket(ticket, service_url):
             # this extra step will guarantee that 2FA are enforced
             # current CAS session created by external login must be cleared first before authentication
             if external_credential:
+                access_token = cas_resp.attributes.get('orcidAccessToken', None)
```
Comment (cslzchen) [inline framework/auth/cas.py:323]: Similarly, let's get refresh token as optional.

File: framework/auth/cas.py [hunk 27]
```diff
@@ -325,6 +350,7 @@ def make_response_from_ticket(ticket, service_url):
             user = {
                 'external_id_provider': external_credential['provider'],
                 'external_id': external_credential['id'],
+                'external_id_access_token': cas_resp.attributes.get('orcidAccessToken', None),
```
Comment (cslzchen) [inline framework/auth/cas.py:359]: Ditto

File: website/settings/defaults.py (hunk 18 above)
Comment (cslzchen) [inline website/settings/defaults.py]: Why do we still need this one?

**Review by cslzchen — CHANGES_REQUESTED** [review]
3rd pass done 🌟 

This is the main PR now, please update ticket title.

File: osf/models/user.py (hunk 20 above)
Comment (Vlad0n20) [reply osf/models/user.py:2158]: users_with_multiple_orcids = OSFUser.objects.filter(external_identity__has_key='ORCID') 
                                                                                          
  count = 0                                                                               
  for user in users_with_multiple_orcids:                                                 
      orcid_ids = user.external_identity.get('ORCID', {})                                 
      if len(orcid_ids) > 1:                                                              
          count += 1                                                                      
          print(user._id, list(orcid_ids.keys()))                                         
                                                                                          
  print(f'{count} users found with more than one ORCID ID')
run with python manage.py osf_shell

File: osf/models/user.py [hunk 28]
```diff
@@ -817,7 +829,12 @@ def merge_user(self, user):
                         self.external_identity[service] = {
                             service_id: status
                         }
+
+                token_entry = user.external_identity_tokens.get(service, {}).get(service_id)
+                if token_entry:
+                    self.external_identity_tokens.setdefault(service, {})[service_id] = token_entry
```
Comment (cslzchen) [inline osf/models/user.py:835]: Not blocking the merge, but I have a question on how external identity is handled during merge?

e.g. two user have different ORCiD, it seems this will create two ORCiD entries, and I think this is the same behavior for tokens

File: osf/models/user.py (hunk 20 above)
Comment (cslzchen) [reply osf/models/user.py:2158]: Per my slack DM: for the cloud request of multiple ORCiD accounts, write a more efficient query using `iterator` with `chunk_size` (we have 2 million users).

File: osf/models/user.py [hunk 29]
```diff
@@ -2137,6 +2154,55 @@ def _clear_identifying_information(self):
...
+        orcid_token = orcid_access_token or orcid_refresh_token
+        sentry.log_message(
+            f'[GDPR delete; _clear_identifying_information] user={self._id}: '
+            f'{"found" if orcid_token else "no"} ORCID token to revoke',
+            level=logging.INFO,
+        )
+        if not (orcid_id and orcid_token):
+            raise UserStateError(
+                'User do not have connected ORCID'
+            )
```
Comment (cslzchen) [inline osf/models/user.py:2173]: Because we will have users who have external identity but no token, let's have different messages returned back to the admin.

Case one: if user has identity but no token for the identity -> error and log sentry
Case two: if user has both identity and tokens matching the identity -> success
Case three: if user has no identity but have token -> error and log sentry

File: framework/auth/__init__.py [hunk 30]
```diff
@@ -60,6 +60,8 @@ def external_first_login_authenticate(user_dict, response):
     data = {
         'auth_user_external_id_provider': user_dict['external_id_provider'],
         'auth_user_external_id': user_dict['external_id'],
+        'auth_user_external_id_access_token': user_dict.get('external_id_access_token'),
+        'auth_user_external_id_refresh_token': user_dict.get('external_id_refresh_token'),
```
Comment (cslzchen) [inline framework/auth/__init__.py:64]: Out of the scope of PR but is fine to have it here.

File: framework/auth/views.py [hunk 31]
```diff
@@ -1050,6 +1050,8 @@ def external_login_email_post():
 
     external_id_provider = session.get('auth_user_external_id_provider', None)
     external_id = session.get('auth_user_external_id', None)
+    external_id_access_token = session.get('auth_user_external_id_access_token', None)
+    external_id_refresh_token = session.get('auth_user_external_id_refresh_token', None)
```
Comment (cslzchen) [inline framework/auth/views.py:1054]: Ditto, out of scope but it's ok

File: framework/auth/views.py [hunk 32]
```diff
@@ -1103,6 +1105,7 @@ def external_login_email_post():
                 user.accepted_terms_of_service = timezone.now()
             # 2. add unconfirmed email and send confirmation email
             user.add_unconfirmed_email(clean_email, external_identity=external_identity)
+            cas.save_orcid_access_token_to_user(user, external_id, external_id_access_token, external_id_refresh_token)
```
Comment (cslzchen) [inline framework/auth/views.py:1108]: Ditto

And another suggestion is it's better for `save_orcid_access_token_to_user` to be in `OSFUser`, but we can fix it in your second PR.

File: framework/auth/views.py [hunk 33]
```diff
@@ -1131,6 +1134,7 @@ def external_login_email_post():
                 campaign=None,
                 accepted_terms_of_service=accepted_terms_of_service
             )
+            cas.save_orcid_access_token_to_user(user, external_id, external_id_access_token, external_id_refresh_token)
```
Comment (cslzchen) [inline framework/auth/views.py:1137]: Ditto

**Review by cslzchen — APPROVED** [review]
Functionality looks good, I am going to merge and deploy to staging1.

As for my review comments, you can include it in your follow-up PR https://github.com/CenterForOpenScience/osf.io/pull/11888

File: framework/auth/cas.py (hunk 1 above)
Comment (Vlad0n20) [reply framework/auth/cas.py:265]: Done — `save_orcid_access_token_to_user` now logs an error to Sentry when the access token is missing, and a separate warning to Sentry when the refresh token is missing (instead of one combined warning).

File: osf/models/user.py (hunk 29 above)
Comment (Vlad0n20) [reply osf/models/user.py:2173]: Done — `_clear_identifying_information` now checks `external_identity` and `external_identity_tokens` independently and handles the three cases:
- Identity but no token → `UserStateError` + error logged to Sentry, no revoke call.
- Identity + matching token → proceeds with the ORCID revoke call (existing behavior).
- Token but no identity → `UserStateError` + error logged to Sentry, no revoke call.

Also added the implicit fourth case (neither identity nor token, i.e. the common case for users who never connected ORCID): skips revoke and continues the delete, since the old code was raising `UserStateError` for every non-ORCID user. Updated/added tests in `TestUserGdprDelete` to cover all four cases.

File: framework/auth/views.py (hunk 32 above)
Comment (Vlad0n20) [reply framework/auth/views.py:1108]: Done — moved `save_orcid_access_token_to_user` from `framework/auth/cas.py` into `OSFUser` (`osf/models/user.py`) as an instance method. Updated the three call sites (`framework/auth/cas.py::make_response_from_ticket`, and the two in `framework/auth/views.py::external_login_email_post`) to call `user.save_orcid_access_token_to_user(...)` instead.

File: osf/models/user.py (hunk 20 above)
Comment (Vlad0n20) [reply osf/models/user.py:2158]: Updated with `.iterator(chunk_size=...)` so it doesn't load the full 2M-row queryset into memory:

```python
users_with_multiple_orcids = OSFUser.objects.filter(external_identity__has_key='ORCID').iterator(chunk_size=2000)

count = 0
for user in users_with_multiple_orcids:
    orcid_ids = user.external_identity.get('ORCID', {})
    if len(orcid_ids) > 1:
        count += 1
        print(user._id, list(orcid_ids.keys()))

print(f'{count} users found with more than one ORCID ID')
```

Run with `python manage.py osf_shell`

Comment (Vlad0n20) [reply osf/models/user.py:2158]: users_with_multiple_orcids=OSFUser.objects.filter(external_identity__has_key='ORCID').iterator(chunk_size=2000)                                                                              
  count = 0
  for user in users_with_multiple_orcids:
      orcid_ids = user.external_identity.get('ORCID', {})
      if len(orcid_ids) > 1:
          count += 1
          print(user._id, list(orcid_ids.keys()))
  print(f'{count} users found with more than one ORCID ID')

File: osf/models/user.py (hunk 28 above)
Comment (Vlad0n20) [reply osf/models/user.py:835]: Checked — confirmed, this is real. Tracing it: if `self` already has ORCID id A `VERIFIED` and the merged-away `user` has a different id B also `VERIFIED`, the `if not (...)` guard is true for B (it's not yet in `self.external_identity['ORCID']`), so it falls into `self.external_identity[service].update({service_id: status})`, which *adds* B alongside A instead of replacing it. Same for the new token line (`setdefault(service, {})[service_id] = token_entry`) — it adds a second key. So a merge can leave `self.external_identity['ORCID']` and `self.external_identity_tokens['ORCID']` each with two entries.

This loop itself is pre-2016 generic multi-provider merge logic (`git blame` points to the original commit), so it's not something this PR introduced — the new token line just mirrors the same existing per-`service_id` loop. But it does mean the "a user has at most one ORCID identity" assumption `_clear_identifying_information` relies on (`next(iter(...))`, grabs one arbitrary entry) doesn't hold for merged accounts: if a user merges in an account with a different verified ORCID, GDPR delete will only revoke one of the two tokens and silently leave the other with live trusted-party access to ORCID.

Agreed this shouldn't block this PR. Want this tracked as a separate ticket, or should I fix `_clear_identifying_information` now to iterate/revoke *all* ORCID token entries instead of just the first one?
